File: forms/workplace-safety-assessment/full-stack-with-loco-tera-htmx-alpine/tests/engine/safety_grader_test.rs
use workplace_safety_assessment_tera_crate::engine::rules::all_rules;
use workplace_safety_assessment_tera_crate::engine::safety_grader::{grade, highest_outcome};
use workplace_safety_assessment_tera_crate::engine::types::*;

/// Build a workplace-safety audit with every "no-is-bad" answered yes and
/// every "yes-is-bad" answered no — a fully compliant site.
fn fully_compliant_audit() -> AssessmentData {
    AssessmentData {
        site_details: SiteDetails {
            auditor_name: "Pat Jones".to_string(),
            auditor_role: "Safety Officer".to_string(),
            audit_date: "2026-06-01".to_string(),
            site_name: "St James's Hospital".to_string(),
            site_address: "Beckett St, Leeds".to_string(),
            department_area: "Ward 4".to_string(),
            site_manager: "Lee Park".to_string(),
            previous_audit_date: "2025-12-01".to_string(),
            previous_findings_closed: "yes".to_string(),
        },
        ppe_hazard_controls: PPEHazardControls {
            ppe_available: "yes".to_string(),
            ppe_correctly_used: "yes".to_string(),
            ppe_stock_maintained: "yes".to_string(),
            hazard_signage_visible: "yes".to_string(),
            signage_legible: "yes".to_string(),
            housekeeping_satisfactory: "yes".to_string(),
            slip_trip_hazards_controlled: "yes".to_string(),
            observations: String::new(),
        },
        chemical_biological_hazards: ChemicalBiologicalHazards {
            coshh_register_present: "yes".to_string(),
            sds_available: "yes".to_string(),
            chemicals_labelled_correctly: "yes".to_string(),
            chemicals_stored_securely: "yes".to_string(),
            spill_kits_available: "yes".to_string(),
            untreated_spills_observed: "no".to_string(),
            sharps_containers_in_date: "yes".to_string(),
            clinical_waste_segregated: "yes".to_string(),
            biological_risk_assessment_current: "yes".to_string(),
            observations: String::new(),
        },
        electrical_safety: ElectricalSafety {
            pat_testing_in_date: "yes".to_string(),
            fixed_wiring_test_in_date: "yes".to_string(),
            damaged_equipment_observed: "no".to_string(),
            overloaded_sockets_observed: "no".to_string(),
            extension_leads_managed_safely: "yes".to_string(),
            consumer_unit_accessible: "yes".to_string(),
            observations: String::new(),
        },
        fire_safety: FireSafety {
            fire_risk_assessment_current: "yes".to_string(),
            fire_extinguishers_serviced: "yes".to_string(),
            fire_extinguishers_accessible: "yes".to_string(),
            fire_alarm_tested_weekly: "yes".to_string(),
            emergency_egress_clear: "yes".to_string(),
            emergency_lighting_functional: "yes".to_string(),
            fire_doors_held_open_illegally: "no".to_string(),
            assembly_point_signposted: "yes".to_string(),
            observations: String::new(),
        },
        ergonomics_manual_handling: ErgonomicsManualHandling {
            manual_handling_assessment_current: "yes".to_string(),
            lifting_aids_available: "yes".to_string(),
            dse_assessments_completed: "yes".to_string(),
            workstations_adjustable: "yes".to_string(),
            repetitive_strain_concerns: "no".to_string(),
            patient_handling_plans_in_place: "yes".to_string(),
            observations: String::new(),
        },
        emergency_procedures: EmergencyProcedures {
            evacuation_procedure_posted: "yes".to_string(),
            first_aid_kits_stocked: "yes".to_string(),
            first_aider_roster_current: "yes".to_string(),
            aed_available: "yes".to_string(),
            aed_service_in_date: "yes".to_string(),
            emergency_contacts_displayed: "yes".to_string(),
            drill_conducted_last12_months: "yes".to_string(),
            observations: String::new(),
        },
        training_competence: TrainingCompetence {
            mandatory_training_up_to_date: "yes".to_string(),
            fire_marshals_trained: "yes".to_string(),
            manual_handling_training_current: "yes".to_string(),
            infection_control_training_current: "yes".to_string(),
            training_records_accessible: "yes".to_string(),
            induction_for_new_starters_completed: "yes".to_string(),
            observations: String::new(),
        },
        incident_reporting: IncidentReporting {
            incident_reporting_system_used: "yes".to_string(),
            riddor_reportable_incidents_reported: "yes".to_string(),
            near_miss_reporting_active: "yes".to_string(),
            incidents_last12_months: Some(2),
            near_misses_last12_months: Some(15),
            lessons_learned_shared: "yes".to_string(),
            actions_from_incidents_tracked: "yes".to_string(),
            observations: String::new(),
        },
        signoff_action_plan: SignoffActionPlan {
            action_items: vec![],
            overall_summary: "Site fully compliant.".to_string(),
            auditor_signature: "Pat Jones".to_string(),
            signoff_date: "2026-06-01".to_string(),
            debrief_delivered: "yes".to_string(),
        },
    }
}

#[test]
fn empty_audit_is_compliant_and_zero_answered() {
    let data = AssessmentData::default();
    let result = grade(&data);
    assert_eq!(result.outcome, "compliant");
    assert_eq!(result.answered_count, 0);
    assert!(result.fired_rules.is_empty());
}

#[test]
fn fully_compliant_audit_grades_compliant() {
    let data = fully_compliant_audit();
    let result = grade(&data);
    assert_eq!(result.outcome, "compliant");
    // Every rule in all_rules() should be answered.
    assert_eq!(result.answered_count as usize, all_rules().len());
}

#[test]
fn coshh_missing_yields_critical_outcome() {
    let mut data = fully_compliant_audit();
    data.chemical_biological_hazards.coshh_register_present = "no".to_string();
    let result = grade(&data);
    assert_eq!(result.outcome, "critical");
    assert!(result.fired_rules.iter().any(|r| r.id == "WS-009" && r.grade == 4));
}

#[test]
fn damaged_electrical_yields_critical_outcome() {
    let mut data = fully_compliant_audit();
    data.electrical_safety.damaged_equipment_observed = "yes".to_string();
    let result = grade(&data);
    assert_eq!(result.outcome, "critical");
    assert!(result.fired_rules.iter().any(|r| r.id == "WS-020" && r.grade == 4));
}

#[test]
fn signage_no_yields_minor_outcome() {
    let mut data = fully_compliant_audit();
    data.ppe_hazard_controls.signage_legible = "no".to_string();
    let result = grade(&data);
    assert_eq!(result.outcome, "minor");
    assert!(result.fired_rules.iter().any(|r| r.id == "WS-006" && r.grade == 2));
}

#[test]
fn slip_trip_no_yields_major_outcome() {
    let mut data = fully_compliant_audit();
    data.ppe_hazard_controls.slip_trip_hazards_controlled = "no".to_string();
    let result = grade(&data);
    assert_eq!(result.outcome, "major");
    assert!(result.fired_rules.iter().any(|r| r.id == "WS-008" && r.grade == 3));
}

#[test]
fn unanswered_items_do_not_fire() {
    let mut data = fully_compliant_audit();
    data.ppe_hazard_controls.ppe_available = String::new();
    let result = grade(&data);
    // No rule fires for the unanswered item; outcome remains compliant.
    assert!(result.fired_rules.iter().all(|r| r.id != "WS-002"));
    assert_eq!(result.outcome, "compliant");
}

#[test]
fn rule_ids_are_unique() {
    let rules = all_rules();
    let mut ids: Vec<&str> = rules.iter().map(|r| r.id).collect();
    ids.sort();
    let len_before = ids.len();
    ids.dedup();
    assert_eq!(ids.len(), len_before, "rule IDs must be unique");
}

#[test]
fn highest_outcome_picks_worst() {
    let fired = vec![
        FiredRule {
            id: "X".to_string(),
            category: "X".to_string(),
            description: "X".to_string(),
            grade: 2,
        },
        FiredRule {
            id: "Y".to_string(),
            category: "Y".to_string(),
            description: "Y".to_string(),
            grade: 4,
        },
    ];
    assert_eq!(highest_outcome(&fired), "critical");
}

File: forms/workplace-safety-assessment/front-end-with-svelte/src/lib/engine/safety-grader.test.ts
import { describe, it, expect } from 'vitest';
import { gradeSafety } from './safety-grader';
import { detectAdditionalFlags } from './flagged-issues';
import { safetyRules } from './safety-rules';
import type { AssessmentData } from './types';

/** A fully-answered, fully-compliant audit (every control in place). */
function createCompliantAudit(): AssessmentData {
	return {
		siteDetails: {
			auditorName: 'Alex Auditor',
			auditorRole: 'Health & Safety Lead',
			auditDate: '2026-06-10',
			siteName: 'Riverside Health Centre',
			siteAddress: '1 Riverside Way',
			departmentArea: 'Outpatients',
			siteManager: 'Sam Manager',
			previousAuditDate: '2025-06-10',
			previousFindingsClosed: 'yes'
		},
		ppeHazardControls: {
			ppeAvailable: 'yes',
			ppeCorrectlyUsed: 'yes',
			ppeStockMaintained: 'yes',
			hazardSignageVisible: 'yes',
			signageLegible: 'yes',
			housekeepingSatisfactory: 'yes',
			slipTripHazardsControlled: 'yes',
			observations: ''
		},
		chemicalBiologicalHazards: {
			coshhRegisterPresent: 'yes',
			sdsAvailable: 'yes',
			chemicalsLabelledCorrectly: 'yes',
			chemicalsStoredSecurely: 'yes',
			spillKitsAvailable: 'yes',
			untreatedSpillsObserved: 'no',
			sharpsContainersInDate: 'yes',
			clinicalWasteSegregated: 'yes',
			biologicalRiskAssessmentCurrent: 'yes',
			observations: ''
		},
		electricalSafety: {
			patTestingInDate: 'yes',
			fixedWiringTestInDate: 'yes',
			damagedEquipmentObserved: 'no',
			overloadedSocketsObserved: 'no',
			extensionLeadsManagedSafely: 'yes',
			consumerUnitAccessible: 'yes',
			observations: ''
		},
		fireSafety: {
			fireRiskAssessmentCurrent: 'yes',
			fireExtinguishersServiced: 'yes',
			fireExtinguishersAccessible: 'yes',
			fireAlarmTestedWeekly: 'yes',
			emergencyEgressClear: 'yes',
			emergencyLightingFunctional: 'yes',
			fireDoorsHeldOpenIllegally: 'no',
			assemblyPointSignposted: 'yes',
			observations: ''
		},
		ergonomicsManualHandling: {
			manualHandlingAssessmentCurrent: 'yes',
			liftingAidsAvailable: 'yes',
			dseAssessmentsCompleted: 'yes',
			workstationsAdjustable: 'yes',
			repetitiveStrainConcerns: 'no',
			patientHandlingPlansInPlace: 'yes',
			observations: ''
		},
		emergencyProcedures: {
			evacuationProcedurePosted: 'yes',
			firstAidKitsStocked: 'yes',
			firstAiderRosterCurrent: 'yes',
			aedAvailable: 'yes',
			aedServiceInDate: 'yes',
			emergencyContactsDisplayed: 'yes',
			drillConductedLast12Months: 'yes',
			observations: ''
		},
		trainingCompetence: {
			mandatoryTrainingUpToDate: 'yes',
			fireMarshalsTrained: 'yes',
			manualHandlingTrainingCurrent: 'yes',
			infectionControlTrainingCurrent: 'yes',
			trainingRecordsAccessible: 'yes',
			inductionForNewStartersCompleted: 'yes',
			observations: ''
		},
		incidentReporting: {
			incidentReportingSystemUsed: 'yes',
			riddorReportableIncidentsReported: 'yes',
			nearMissReportingActive: 'yes',
			incidentsLast12Months: 1,
			nearMissesLast12Months: 12,
			lessonsLearnedShared: 'yes',
			actionsFromIncidentsTracked: 'yes',
			observations: ''
		},
		signoffActionPlan: {
			actionItems: [],
			overallSummary: '',
			auditorSignature: '',
			signoffDate: '',
			debriefDelivered: 'yes'
		}
	};
}

describe('Workplace Safety Grading Engine', () => {
	it('returns compliant for a blank (unanswered) audit', () => {
		const blank = createCompliantAudit();
		// Wipe every Yes/No/N/A answer back to unanswered.
		for (const rule of safetyRules) void rule; // keep import used
		const empty: AssessmentData = JSON.parse(JSON.stringify(blank));
		(Object.keys(empty) as (keyof AssessmentData)[]).forEach((section) => {
			const sec = empty[section] as unknown as Record<string, unknown>;
			for (const key of Object.keys(sec)) {
				if (typeof sec[key] === 'string') sec[key] = '';
			}
		});
		const result = gradeSafety(empty);
		expect(result.outcome).toBe('compliant');
		expect(result.answeredCount).toBe(0);
	});

	it('returns compliant for a fully-compliant audit (all controls in place)', () => {
		const result = gradeSafety(createCompliantAudit());
		expect(result.outcome).toBe('compliant');
		expect(result.answeredCount).toBe(safetyRules.length);
		expect(result.firedRules.every((r) => r.grade === 1)).toBe(true);
	});

	it('returns minor when only grade-2 findings are present', () => {
		const data = createCompliantAudit();
		data.ppeHazardControls.ppeCorrectlyUsed = 'no'; // severity 2
		const result = gradeSafety(data);
		expect(result.outcome).toBe('minor');
	});

	it('returns major when a grade-3 finding is present (no critical)', () => {
		const data = createCompliantAudit();
		data.fireSafety.fireRiskAssessmentCurrent = 'no'; // severity 3
		const result = gradeSafety(data);
		expect(result.outcome).toBe('major');
	});

	it('returns critical when a grade-4 finding is present', () => {
		const data = createCompliantAudit();
		data.chemicalBiologicalHazards.coshhRegisterPresent = 'no'; // severity 4
		const result = gradeSafety(data);
		expect(result.outcome).toBe('critical');
	});

	it('tallies findings by category', () => {
		const data = createCompliantAudit();
		data.fireSafety.emergencyEgressClear = 'no'; // critical
		data.fireSafety.assemblyPointSignposted = 'no'; // minor
		const result = gradeSafety(data);
		const fire = result.findingsByCategory['Fire Safety'];
		expect(fire.critical).toBe(1);
		expect(fire.minor).toBe(1);
		expect(fire.total).toBe(8);
	});

	it('has unique rule IDs', () => {
		const ids = safetyRules.map((r) => r.id);
		expect(new Set(ids).size).toBe(ids.length);
	});
});

describe('Workplace Safety Flagged Issues Detection', () => {
	it('returns no flags for a fully-compliant audit', () => {
		const flags = detectAdditionalFlags(createCompliantAudit());
		expect(flags).toHaveLength(0);
	});

	it('flags missing fire-extinguisher servicing (high)', () => {
		const data = createCompliantAudit();
		data.fireSafety.fireExtinguishersServiced = 'no';
		const flags = detectAdditionalFlags(data);
		expect(flags.some((f) => f.id === 'FLAG-FIRE-001')).toBe(true);
	});

	it('flags missing COSHH register (high)', () => {
		const data = createCompliantAudit();
		data.chemicalBiologicalHazards.coshhRegisterPresent = 'no';
		const flags = detectAdditionalFlags(data);
		expect(flags.some((f) => f.id === 'FLAG-CHEM-002')).toBe(true);
	});

	it('flags unreported RIDDOR incidents (high)', () => {
		const data = createCompliantAudit();
		data.incidentReporting.riddorReportableIncidentsReported = 'no';
		const flags = detectAdditionalFlags(data);
		expect(flags.some((f) => f.id === 'FLAG-INC-001')).toBe(true);
	});

	it('flags near-miss under-reporting when near misses < incidents', () => {
		const data = createCompliantAudit();
		data.incidentReporting.incidentsLast12Months = 5;
		data.incidentReporting.nearMissesLast12Months = 1;
		const flags = detectAdditionalFlags(data);
		expect(flags.some((f) => f.id === 'FLAG-INC-003')).toBe(true);
	});

	it('sorts flags by priority (high first)', () => {
		const data = createCompliantAudit();
		data.fireSafety.fireExtinguishersServiced = 'no'; // high
		data.ppeHazardControls.housekeepingSatisfactory = 'no'; // low
		const flags = detectAdditionalFlags(data);
		const order = { urgent: 0, high: 1, medium: 2, low: 3 } as const;
		const priorities = flags.map((f) => f.priority);
		const sorted = [...priorities].sort((a, b) => order[a] - order[b]);
		expect(priorities).toEqual(sorted);
	});
});
